### ISSUE TITLE
PollingFileSystemWatcher

### ISSUE THREAD

**Issue by janborup** [aelassas/Wexflow#225]
Now you have implemented the PollingFileSystemWatcher class, why not use us it for a new TaskItem entry, that vil expand FilesExists.

Today FileExists is only True/False and has to be Periodic/CRON.
Why not Wait for file created/change/delete then Do... or a While File .....

/Cheers.

**Comment by aelassas** [comment]
@janborup For this, I think that I will add a new task named FileSystemWatcher. 

When a workflow reaches FileSystemWatcher task, it will stay there until it is stopped manually.

Below is a prototype workflow using this task. Tell me what do you think about this.

```xml
<Workflow xmlns="urn:wexflow-schema" id="42" name="Workflow_FileSystemWatcher" description="Workflow_FileSystemWatcher">
	<Settings>
		<Setting name="launchType" value="trigger" />
		<Setting name="enabled" value="true" />
	</Settings>
	<Tasks>
		<Task id="1" name="FileSystemWatcher" description="Checking files" enabled="true">
			<Setting name="folder" value="C:\WexflowTesting\HotFolder" />		
			<Setting name="filter" value="*.*" />
			<Setting name="includeSubFolders" value="false" />
			<Setting name="onFileCreated" value="2, 3" /> <!-- value contains task Ids -->
			<Setting name="onFileChanged" value="2, 3" />
			<Setting name="onFileDeleted" value="4, 5" />
		</Task>
		<Task id="2" name="FilesCopier" description="Copying files" enabled="true">
		  <Setting name="selectFiles" value="1" />
		  <Setting name="destFolder" value="C:\WexflowTesting\SomeFolder" />
		  <Setting name="overwrite" value="true" />
		</Task>
		<Task id="3" name="Md5" description="Generating MD5 sums" enabled="true">
			<Setting name="selectFiles" value="1" />
		</Task>
		<Task id="4" name="FilesLoader" description="Loading mails" enabled="true">
		  <Setting name="file" value="C:\WexflowTesting\Mails\mails.xml" />
		</Task>
		<Task id="5" name="MailsSender" description="Sending mails" enabled="true">
		  <Setting name="selectFiles" value="4" />
		  <Setting name="host" value="smtp.gmail.com" />
		  <Setting name="port" value="587" />
		  <Setting name="enableSsl" value="true" />
		  <Setting name="user" value="username" />
		  <Setting name="password" value="password" />
		</Task>
	</Tasks>
</Workflow>
```

**Comment by janborup** [comment]
@aelassas Absolutely fine.

**Easy to understand, and easy to maintain.**

I think many of my future Workflows will depend on this FileSystemWatcher task, and maybe this will be a kickstarter for starting sub-workflows

<Workflow xmlns="urn:wexflow-schema" id="42" name="Workflow_FileSystemWatcher" description="Workflow_FileSystemWatcher">
	<Settings>
		<Setting name="launchType" value="startup" />
		<Setting name="enabled" value="true" />
	</Settings>
	<Tasks>
		<Task id="1" name="FileSystemWatcher" description="Checking files" enabled="true">
			<Setting name="folder" value="C:\WexflowTesting\HotFolder" />		
			<Setting name="filter" value="*.*" />
			<Setting name="includeSubFolders" value="false" />
			<Setting name="onFileCreated" value="2" /> 
			<Setting name="onFileChanged" value="3" />
			<Setting name="onFileDeleted" value="" />
		</Task>
		<Task id="2" name="Workflow" description="Starting workflow 43" enabled="true">
			<Setting name="wexflowWebServiceUri" value="http://localhost:8000/wexflow/" />
			<Setting name="username" value="admin" />
			<Setting name="password" value="wexflow2018" />
			<Setting name="action" value="start" />
			<Setting name="id" value="43" />
		</Task>
		<Task id="3" name="Workflow" description="Starting workflow 10 and 11" enabled="true">
			<Setting name="wexflowWebServiceUri" value="http://localhost:8000/wexflow/" />
			<Setting name="username" value="admin" />
			<Setting name="password" value="wexflow2018" />
			<Setting name="action" value="start" />
			<Setting name="id" value="10" />
			<Setting name="id" value="11" />
		</Task>
	</Tasks>
</Workflow>

**Comment by aelassas** [comment]
@janborup Great! Thanks for your feadback. I will work on this new task as soon as possible.

**Comment by aelassas** [comment]
@janborup I have just finished [FileSystemWatcher](https://github.com/aelassas/Wexflow/wiki/FileSystemWatcher) task. Can you please pull the latest version and test it please. [Here](https://github.com/aelassas/Wexflow/blob/master/samples/dotnet-core/windows/Wexflow/Workflows/Workflow_FileSystemWatcher.xml) is a sample workflow for .net core. You must create HotFolder, Folder1 and Folder2 in C:\WexflowTesting.

**Comment by janborup** [comment]
Yes.... I followed your work, and I will give it at test try. Thanks.

This task is a great asset for Wexflow 😁😎

**Comment by janborup** [comment]
**Test several steps with this task in .Net Core, and everything is working smooth 😋😎**

I only found 2 little funny issues:
I have changed the workflow-type to "startup", and when I stop the server with Ctrl+C, the server exits with this error: 
> 2020-04-26 21:46:56,110 ERROR [1] - An error occured while stopping the workflow : {id: 151, name: Workflow_FileSystemWatcher, enabled: True, launchType: Startup}
> System.PlatformNotSupportedException: Thread abort is not supported on this platform.
>    at System.Threading.Thread.Abort()
>    at Wexflow.Core.Workflow.Stop() in C:\Work\WexFlow\src\dotnet-core\Wexflow.Core\Workflow.cs:line 157

Another not related problem is the last days the Wexflow DashBoard is repeating the workflows, but don't know if it means anything.
![image](https://user-images.githubusercontent.com/6866443/80318151-45c6e500-8808-11ea-99eb-3c162a70269e.png)

In before only one instance task showed up in the dashboard, even if the tasks have been runned several times.

/Cheers

**Comment by janborup** [comment]
Another related issue.
It the workflow-type is "startup", you cannot stop/start the job.

![image](https://user-images.githubusercontent.com/6866443/80318247-254b5a80-8809-11ea-88a3-d695aaa1329a.png)

**Comment by aelassas** [comment]
@janborup For duplicated entries on the dashboard on startup, I fixed the issue. For stopping FileSystemWatcher workflow, it is also fixed. I added the ability to stop workflows in .net core

Can you pull the latest version and test again please?

**Comment by janborup** [comment]
Nope.... Now something else is broke in the .Net Core version.
The serser won't start...

![image](https://user-images.githubusercontent.com/6866443/80322361-d6abb980-8824-11ea-98ce-0c0dc9320ffc.png)

**Comment by aelassas** [comment]
@janborup I usually get this error when the port 8000 is used by another instance of Wexflow. Check that the port 8000 is free and try again. Otherwise, reboot your computer and try again. I don't get this error on my side.

**Comment by janborup** [comment]
@aelassas Yes... I had a conflict with a double Wexflow Service installation, so that's why the socket was occupied.

Just testet the rest of the FileWatcher, also with multiple Watches, and this also works fine.

Below is a very good example, on Watching a folder, and sending a notification on change.
```
<Workflow xmlns="urn:wexflow-schema" id="200" name="HotTemp Folder - Email" description="Send email at changes in Temp folder">
	<Settings>
		<Setting name="launchType" value="startup" />
		<Setting name="enabled" value="true" />
		<Setting name="approval" value="false" />
		<Setting name="enableParallelJobs" value="true" />
	</Settings>
	<LocalVariables />
	<Tasks>
		<Task id="1" name="FileSystemWatcher" description="Watching C:\Temp" enabled="true">
			<Setting name="folderToWatch" value="c:\temp" />
			<Setting name="onFileCreated" value="2,3" />
			<Setting name="onFileChanged" value="2,3" />
		</Task>
		<Task id="2" name="FilesLoader" description="Get email template" enabled="true">
			<Setting name="file" value="C:\WexflowTesting\Mails\mails.xml" />
		</Task>
		<Task id="3" name="MailsSender" description="Sending mail" enabled="true">
			<Setting name="selectAttachments" value="1" />
			<Setting name="selectFiles" value="2" />
			<Setting name="host" value="smtp.gmail.com" />
			<Setting name="port" value="587" />
			<Setting name="enableSsl" value="true" />
			<Setting name="user" value="john@doe.dk" />
			<Setting name="password" value="secret-password" />
		</Task>
	</Tasks>
</Workflow>

```
Case closed - Good job. 👍

**Comment by janborup** [comment]
A little important enhancement to this.

Would you please implement at task-setting:
<Setting name="onWatchStarted" value="2,3" />

The problem is, when the WexFlow has been down or first time you start WexFlow, and there is new files i a folder, the Watch does not do anything, so if there is files in the Watch folder, do this task..

Thanks.

**Comment by aelassas** [comment]
@janborup Do you want me to add onFileFound event? This event will trigger tasks on each file found in the hot folder before starting to watch the hot folder.

**Comment by janborup** [comment]
Yes. Very fine🏌️‍♀️

Med venlig hilsen

Jan Borup Coyle

**Comment by aelassas** [comment]
@janborup I added onFileFound event to FileSystemWatcher for both .net and .net core. This event is triggered on file found before starting watch process. Can you pull the latest version and test it please?

**Comment by aelassas** [comment]
@janborup Can I close this issue?
